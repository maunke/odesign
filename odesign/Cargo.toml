[package]
name = "odesign"
version.workspace = true
edition.workspace = true
description.workspace = true
repository.workspace = true
keywords.workspace = true
categories.workspace = true
license.workspace = true

[dependencies]
nalgebra = { workspace = true }
num-dual = { workspace = true }
faer = { workspace = true }
faer-ext = { workspace = true }
rayon = "1.10.0"
thiserror = { workspace = true }
aquamarine = { workspace = true }
katexit = { workspace = true }
odesign_derive = { path = "../odesign-derive"}

[dev-dependencies]
rand = "0.9.0"
trybuild = "1.0"
