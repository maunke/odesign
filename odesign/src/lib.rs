#![forbid(unsafe_code)]
#![warn(missing_docs)]
#![doc = include_str!("../README.md")]
mod error;
mod feature;
mod grid;
mod interior_point_method;
mod linear_model;
mod optimal_design;
mod optimality;
mod utils;

pub use error::{Error, Result};
pub use feature::{Feature, FeatureFunction, FeatureSet};
pub use grid::Grid;
pub use interior_point_method::{
    LinearEqualityConstraint, NLPFunctionTarget, NLPSolver, NLPSolverConstraints, NLPSolverOptions,
};
pub use linear_model::LinearModel;
pub use odesign_derive::Feature;
pub use optimal_design::{
    CustomDesignBound, Design, DesignBound, DesignConstraint, OptimalDesign, OptimalDesignCriteria,
};
pub use optimality::{
    AOptimality, COptimality, DOptimality, Optimalities, Optimality, OptimalityMeasures,
};
pub use utils::{IntoSVector, MatrixDRows};
