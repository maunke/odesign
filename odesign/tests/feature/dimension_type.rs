use nalgebra::SVector;
use num_dual::DualNum;
use odesign::{Feature, FeatureFunction};

#[derive(Feature)]
#[dimension = "1"]
struct Monomial {
    i: i32,
    j: i32,
}

impl FeatureFunction<2> for Monomial {
    fn f<D: DualNum<f64>>(&self, x: &SVector<D, 2>) -> D {
        x[0].powi(self.i) * x[1].powi(self.j)
    }
}

fn main() {
    Monomial { i: 0, j: 0 };
}
