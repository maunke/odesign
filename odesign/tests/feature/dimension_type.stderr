error: Missing #[dimension = <d>] attribute where d of type usize is equal to input dimension of feature function
 --> tests/feature/dimension_type.rs:7:8
  |
7 | struct Monomial {
  |        ^^^^^^^^
